// builderoo/hooks/simple.ts
import {useContext} from "react";
import {Blocks, Config, Props} from "../schema";
import {AppContext, CartContext, PageSpecificModelDataContext, SectionDataContext, UserContext,
  BlockContextInternal, EditorContextInternal, SectionContextInternal
} from "../context";


export const useApp = () => useContext(AppContext)
export const useMenu = () => useContext(AppContext).menu || {main: [], footer: []}
export const useIsDevelopMode = () => useContext(AppContext).isDevelopMode
export const useIsEditorMode = () => useContext(AppContext).isEditorMode
export const useUser = () => useContext(UserContext).user || null
export const useCart = () => useContext(CartContext).cart || {items: []}
export const __useInternalExtraLinks = () => useContext(AppContext).extraLinks ?? {}

export function useConfig<T = {[key: string]: any}>(): Config<T>{
  return useContext(AppContext).config || {}
}

export function useProps<T = {[key: string]: any}>(): Props<T>{
  return useContext(SectionDataContext).props || {}
}

export function useBlocks<T>(): Blocks<T>[]{
  return useContext(SectionDataContext).blocks || []
}

export const useProduct = () => useContext(PageSpecificModelDataContext).product
export const useCollection = () => useContext(PageSpecificModelDataContext).collection
export const useCollectionList = () => useContext(PageSpecificModelDataContext).collections
export const useArticle = () => useContext(PageSpecificModelDataContext).article
export const usePage = () => useContext(PageSpecificModelDataContext).page
export const useOrder = () => useContext(PageSpecificModelDataContext).order
export const useOrderList = () => useContext(PageSpecificModelDataContext).orders || []


export const useSectionContextInternal = () => useContext(SectionContextInternal);
export const useBlockInternal = () => useContext(BlockContextInternal)

export const useEditorContextInternal= () => useContext(EditorContextInternal);
export const useIsInEditMode = () => useContext(EditorContextInternal) != null
